File: frontend/lib/api/delivery-service.ts
/**
 * Service de délégation à un agent de livraison (coding agent) - TF-AGENT-DELIVERY.
 * Slice 2 : liste des providers délégables (pour le picker « déléguer à... »).
 */

import { apiClient } from "./client";
import { DELIVERY_ROUTES } from "../config/api-routes";

/** Provider d'agent de livraison exposé au picker. `available=false` = affiché « à venir ». */
export interface DeliveryProvider {
  key: string;
  displayName: string;
  logoKey: string;
  available: boolean;
  models: string[];
}

/** Liste les providers délégables du workspace (Claude Code / Copilot / Cursor...). */
export async function getDeliveryProviders(slug: string): Promise<DeliveryProvider[]> {
  const res = await apiClient.get<{ data: DeliveryProvider[] }>(DELIVERY_ROUTES.PROVIDERS(slug));
  return res.data.data;
}

export type DeliveryRunStatus = "QUEUED" | "RUNNING" | "DONE" | "FAILED";

/** État d'un run de délégation d'une issue. `summary`/`resultUrl` non nuls quand DONE. */
export interface DeliveryRun {
  id: number;
  issueId: number;
  providerKey: string;
  model: string | null;
  status: DeliveryRunStatus;
  summary: string | null;
  resultUrl: string | null;
  error: string | null;
  startedById: number | null;
  createdAt: string | null;
  updatedAt: string | null;
}

/** Délègue une issue à un agent (crée un run, exécuté en arrière-plan côté serveur). */
export async function delegateIssue(
  slug: string,
  issueId: number,
  providerKey: string,
  model?: string,
): Promise<DeliveryRun> {
  const res = await apiClient.post<{ data: DeliveryRun }>(
    DELIVERY_ROUTES.DELEGATE(slug, issueId),
    { providerKey, model },
  );
  return res.data.data;
}

/** Dernier run de délégation d'une issue, ou null si aucune délégation. */
export async function getIssueRun(slug: string, issueId: number): Promise<DeliveryRun | null> {
  const res = await apiClient.get<{ data: DeliveryRun | null }>(DELIVERY_ROUTES.RUN(slug, issueId));
  return res.data.data;
}

/** Providers de délégation dont l'exécution dépend d'une clé API stockée par workspace. */
export type DeliveryKeyProvider = "anthropic" | "cursor";

/** État de la connexion d'une clé API de délégation (Anthropic, Cursor... TF-AGENT-DELIVERY). */
export interface DeliveryKeyStatus {
  /** Une clé est connectée (délégation vers ce provider possible). */
  connected: boolean;
  /** Indice non sensible pour reconnaître la clé (4 derniers caractères, ex. "...AB12"), ou null. */
  keyHint: string | null;
}

/** Lit l'état d'une clé de délégation du workspace (la clé n'est jamais renvoyée en clair). */
export async function getDeliveryKey(slug: string, provider: DeliveryKeyProvider): Promise<DeliveryKeyStatus> {
  const res = await apiClient.get<{ data: DeliveryKeyStatus }>(DELIVERY_ROUTES.KEY(slug, provider));
  return res.data.data;
}

/** Connecte (ou remplace) une clé de délégation du workspace (OWNER/ADMIN). */
export async function connectDeliveryKey(
  slug: string,
  provider: DeliveryKeyProvider,
  apiKey: string,
): Promise<DeliveryKeyStatus> {
  const res = await apiClient.post<{ data: DeliveryKeyStatus }>(DELIVERY_ROUTES.KEY(slug, provider), { apiKey });
  return res.data.data;
}

/** Déconnecte une clé de délégation du workspace (OWNER/ADMIN). */
export async function disconnectDeliveryKey(slug: string, provider: DeliveryKeyProvider): Promise<void> {
  await apiClient.delete(DELIVERY_ROUTES.KEY(slug, provider));
}

File: backend/tf-api/src/main/java/com/taskforce/tf_api/core/service/delivery/DeliveryService.java
package com.taskforce.tf_api.core.service.delivery;

import java.util.Optional;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.taskforce.tf_api.core.dto.response.AnthropicStatusResponse;
import com.taskforce.tf_api.core.dto.response.DeliveryRunResponse;
import com.taskforce.tf_api.core.enums.DeliveryRunStatus;
import com.taskforce.tf_api.core.enums.IntegrationProvider;
import com.taskforce.tf_api.core.model.DeliveryRun;
import com.taskforce.tf_api.core.model.Integration;
import com.taskforce.tf_api.core.model.Issue;
import com.taskforce.tf_api.core.model.User;
import com.taskforce.tf_api.core.model.Workspace;
import com.taskforce.tf_api.core.repository.DeliveryRunRepository;
import com.taskforce.tf_api.core.repository.IntegrationRepository;
import com.taskforce.tf_api.core.repository.IssueRepository;
import com.taskforce.tf_api.core.repository.UserRepository;
import com.taskforce.tf_api.core.service.ProjectVisibilityGuard;
import com.taskforce.tf_api.core.service.brain.BrainAccessGuard;
import com.taskforce.tf_api.shared.exception.BusinessException;
import com.taskforce.tf_api.shared.exception.ResourceNotFoundException;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

/**
 * Orchestration de la délégation d'une tâche à un agent de livraison (TF-AGENT-DELIVERY slice 3).
 *
 * <p>Comme {@code AnalysisJobService} : cette classe crée/lit le run dans sa transaction mais ne
 * déclenche PAS le runner (@Async) elle-même — c'est le contrôleur qui lance {@link DeliveryRunner}
 * une fois le run committé, sinon le thread async pourrait démarrer avant le commit et ne pas voir
 * la ligne.</p>
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class DeliveryService {

    private final DeliveryRunRepository runRepository;
    private final IssueRepository issueRepository;
    private final UserRepository userRepository;
    private final DeliveryAgentProviderRegistry registry;
    private final ProjectVisibilityGuard visibilityGuard;
    private final BrainAccessGuard access;
    private final IntegrationRepository integrationRepository;

    /**
     * Crée un run de délégation (QUEUED) pour une issue vers un provider disponible. Écriture sur
     * l'issue (LEAD/ADMIN/OWNER). Le contrôleur déclenche l'exécution après le commit.
     */
    @Transactional
    public DeliveryRunResponse delegate(String slug, Long issueId, String providerKey, String model, Long userId) {
        Issue issue = scopedIssue(slug, issueId);
        visibilityGuard.assertCanWrite(issue.getProject(), userId);

        DeliveryAgentProvider provider = registry.get(providerKey);
        if (provider == null) {
            throw new BusinessException("Provider inconnu : " + providerKey);
        }
        if (!provider.available()) {
            throw new BusinessException("Ce provider n'est pas encore disponible : " + providerKey);
        }

        String resolvedModel = (model != null && !model.isBlank())
            ? model.trim()
            : (provider.models().isEmpty() ? null : provider.models().get(0));

        User startedBy = userRepository.findById(userId).orElse(null);
        DeliveryRun run = DeliveryRun.builder()
            .issue(issue)
            .providerKey(provider.key())
            .model(resolvedModel)
            .status(DeliveryRunStatus.QUEUED)
            .startedBy(startedBy)
            .build();
        run = runRepository.save(run);

        log.info("Délégation issue {} -> provider {} (run {}, QUEUED)", issueId, provider.key(), run.getId());
        return toResponse(run);
    }

    /** Dernier run de délégation d'une issue (ou vide), pour afficher l'état courant. */
    @Transactional(readOnly = true)
    public Optional<DeliveryRunResponse> latestRun(String slug, Long issueId, Long userId) {
        Issue issue = scopedIssue(slug, issueId);
        visibilityGuard.assertCanView(issue.getProject(), userId);
        return runRepository.findTopByIssueIdOrderByCreatedAtDesc(issue.getId()).map(this::toResponse);
    }

    // =========================================================================
    // Clé API Anthropic du workspace (délégation Claude via l'API, B1)
    // =========================================================================

    /**
     * Connecte (ou remplace) la clé API Anthropic du workspace. Réservé OWNER/ADMIN (secret d'espace,
     * cf. RBAC intégrations). La clé est stockée <b>chiffrée</b> ({@code Integration.accessToken}) et
     * n'est jamais renvoyée en clair - elle est vérifiée à la première délégation Claude.
     */
    @Transactional
    public AnthropicStatusResponse connectAnthropic(String slug, Long userId, String apiKey) {
        Workspace ws = access.resolveAndAuthorizeOwner(slug, userId);
        Integration integ = integrationRepository
            .findByWorkspaceIdAndProvider(ws.getId(), IntegrationProvider.ANTHROPIC)
            .orElseGet(Integration::new);
        integ.setWorkspace(ws);
        integ.setProvider(IntegrationProvider.ANTHROPIC);
        integ.setAccessToken(apiKey.trim());
        integ.setInstalledBy(userRepository.findById(userId).orElse(null));
        integrationRepository.save(integ);
        log.info("Cle Anthropic connectee au workspace {} (delegation Claude)", ws.getId());
        return status(ws);
    }

    /** État de la connexion Anthropic (tout membre du workspace). */
    @Transactional(readOnly = true)
    public AnthropicStatusResponse anthropicStatus(String slug, Long userId) {
        return status(access.resolveAndAuthorize(slug, userId));
    }

    /** Déconnecte la clé Anthropic du workspace. Réservé OWNER/ADMIN. */
    @Transactional
    public void disconnectAnthropic(String slug, Long userId) {
        Workspace ws = access.resolveAndAuthorizeOwner(slug, userId);
        integrationRepository.deleteByWorkspaceIdAndProvider(ws.getId(), IntegrationProvider.ANTHROPIC);
        log.info("Cle Anthropic deconnectee du workspace {}", ws.getId());
    }

    private AnthropicStatusResponse status(Workspace ws) {
        return integrationRepository.findByWorkspaceIdAndProvider(ws.getId(), IntegrationProvider.ANTHROPIC)
            .map(i -> new AnthropicStatusResponse(true, keyHint(i.getAccessToken())))
            .orElseGet(() -> new AnthropicStatusResponse(false, null));
    }

    /** Indice non sensible : les 4 derniers caractères seulement (jamais la clé entière). */
    private String keyHint(String key) {
        return (key == null || key.length() < 4) ? null : "..." + key.substring(key.length() - 4);
    }

    /** Mappe le run en DTO DANS la transaction (les accès paresseux issue/startedBy y sont sûrs). */
    private DeliveryRunResponse toResponse(DeliveryRun r) {
        return new DeliveryRunResponse(
            r.getId(),
            r.getIssue().getId(),
            r.getProviderKey(),
            r.getModel(),
            r.getStatus().name(),
            r.getSummary(),
            r.getResultUrl(),
            r.getError(),
            r.getStartedBy() != null ? r.getStartedBy().getId() : null,
            r.getCreatedAt(),
            r.getUpdatedAt());
    }

    /** Charge une issue en la scopant au workspace du chemin (cross-tenant -> 404, pas 403). */
    private Issue scopedIssue(String slug, Long issueId) {
        Issue issue = issueRepository.findById(issueId)
            .orElseThrow(() -> new ResourceNotFoundException("Issue introuvable"));
        if (issue.getProject() == null || issue.getProject().getWorkspace() == null
            || !slug.equals(issue.getProject().getWorkspace().getSlug())) {
            throw new ResourceNotFoundException("Issue introuvable");
        }
        return issue;
    }
}
